Engine p2p sync test PR

ImTei opened the pull request.

ImTei committed:
Enable EngineQueue to trigger engine p2p sync

ImTei committed:
Add EngineP2PEnabled flag and save into sync.Config

ImTei committed:
Try safe attributes only if engine is synced

ImTei committed:
Allow ACCEPTED of NewPayloadV1 when applying unsafe payload

ImTei committed:
Rename unsafeSyncTarget to engineSyncTarget

ImTei committed:
Add EngineSyncTarget to syncStatus

ImTei committed:
Add e2e test cases for unsafe sync and engine p2p sync

ImTei committed:
Check reorg depth with new unsafe head during FineL2Heads

ImTei committed:
Add SkipSanityCheck feature

ImTei committed:
Fix checking reorg depth to use L1 origin number

ImTei committed:
Make status checker methods for readability

ImTei committed:
Rename engine p2p enabled flag

ImTei committed:
Rename engine p2p enabled flag

ImTei committed:
Fix forkchoiceUpdated status condition

ImTei committed:
Fix happy-path sync spec

pcw109550 commented on op-node/flags/flags.go:
```diff
@@ -208,6 +208,20 @@ var (
 		Required: false,
 	}
+	L2EngineP2PEnabled = &cli.BoolFlag{
+		Name:     "l2.engine-p2p-enabled",
```
I think `l2.engine-p2p.enabled` is better. For consistency with `metrics.enabled`, `pprof.enabled`, `sequencer.enabled`, `heartbeat.enabled`.

There is other alternative: `l2.enable-engine-p2p`

pcw109550 commented on op-node/rollup/sync/config.go:
```diff
@@ -0,0 +1,8 @@
+package sync
+
+type Config struct {
+	// EngineP2PEnabled is true when the EngineQueue can trigger execution engine P2P sync.
+	EngineP2PEnabled bool `json:"engine_p2p_enabled"`
+	// SkipSanityCheck is true when the EngineQueue does not do sanity check on pipeline reset.
+	SkipSanityCheck bool `json:"skip_sanity_check"`
```
Apply lower camel case for consistency.

```suggestion
	// EngineP2PEnabled is true when the EngineQueue can trigger execution engine P2P sync.
	EngineP2PEnabled bool `json:"engineP2PEnabled"`
	// SkipSanityCheck is true when the EngineQueue does not do sanity check on pipeline reset.
	SkipSanityCheck bool `json:"skipSanityCheck"`
```

pcw109550 commented on op-e2e/actions/sync_test.go:
```diff
@@ -92,3 +95,72 @@ func TestFinalizeWhileSyncing(gt *testing.T) {
+		verifier.ActL2PipelineFull(t)
+		// Verifier must advance its unsafe head and engine sync target.
+		require.Equal(t, sequencer.L2Unsafe().Hash, verifier.L2Unsafe().Hash)
+		require.Equal(t, verifier.L2Unsafe().Hash, verifier.EngineSyncTarget().Hash)
```
```suggestion
		// Check engine sync target updated.
		require.Equal(t, sequencer.L2Unsafe().Hash, sequencer.EngineSyncTarget().Hash)
		require.Equal(t, verifier.L2Unsafe().Hash, verifier.EngineSyncTarget().Hash)
```

pcw109550 commented on op-e2e/actions/sync_test.go:
```diff
@@ -92,3 +95,72 @@ func TestFinalizeWhileSyncing(gt *testing.T) {
+		require.NotEqual(t, verifier.L2Unsafe().Hash, verifier.EngineSyncTarget().Hash)
+		require.Equal(t, verifier.L2Unsafe().Hash, verifierUnsafeHead.Hash)
+		require.Equal(t, sequencer.L2Unsafe().Hash, verifier.EngineSyncTarget().Hash)
+	}
```
```suggestion
	for i := 0; i < 10; i++ {
		// Build a L2 block.
		sequencer.ActL2StartBlock(t)
		sequencer.ActL2EndBlock(t)
	}
	// Notify most recent unsafe L2 block to verifier by unsafe gossip
	seqHead, err := seqEngCl.PayloadByLabel(t.Ctx(), eth.Unsafe)
	require.NoError(t, err)
	verifier.ActL2UnsafeGossipReceive(seqHead)(t)
	// Handle unsafe payload
	verifier.ActL2PipelineFull(t)
	// Verifier cannot advance unsafe head
	require.NotEqual(t, sequencer.L2Unsafe().Hash, verifier.L2Unsafe().Hash)
	require.Equal(t, verifier.L2Unsafe().Hash, verifierUnsafeHead.Hash)
	// Verifier must advance only engine sync target
	require.NotEqual(t, verifier.L2Unsafe().Hash, verifier.EngineSyncTarget().Hash)
	require.Equal(t, sequencer.L2Unsafe().Hash, verifier.EngineSyncTarget().Hash)
```

ImTei committed:
Add reorg depth test

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -162,6 +169,7 @@ func (eq *EngineQueue) SystemConfig() eth.SystemConfig {
 
 func (eq *EngineQueue) SetUnsafeHead(head eth.L2BlockRef) {
 	eq.unsafeHead = head
+	eq.engineSyncTarget = head
```
I think we need separate function for this.
```go
func (eq *EngineQueue) SetEngineSyncTarget(head eth.L2BlockRef) {
	eq.engineSyncTarget = head
}
```

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -221,11 +229,26 @@ func (eq *EngineQueue) SafeL2Head() eth.L2BlockRef {
+
+// Determine if the engine is synced to the notified unsafe block.
+// If it returns false, the engine status is SYNCING.
+func (eq *EngineQueue) isUnsafeSynced() bool {
```
This method checks unsafe block finally reached engine sync target. Is there a better name for this, like `isUnsafeReachedEngineSyncTarget()` ?

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -221,11 +229,26 @@ func (eq *EngineQueue) SafeL2Head() eth.L2BlockRef {
+	if eq.unsafePayloads.Len() > 0 {
+		return eq.tryNextUnsafePayload(ctx)
+	}
+	if eq.safeAttributes != nil && eq.isUnsafeSynced() {
```
```suggestion
        if eq.isUnsafeSynced() {
            // Make pipeline first focus to sync unsafe block to engineSyncTarget
            return nil
        }
	if eq.safeAttributes != nil {
```

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -412,6 +431,26 @@ func (eq *EngineQueue) tryUpdateEngine(ctx context.Context) error {
+// It returns true if the status is acceptable.
+func (eq *EngineQueue) checkNewPayloadStatus(status eth.ExecutePayloadStatus) bool {
+	if eq.syncCfg.EngineP2PEnabled {
+		// Allow ACCEPTED and SYNCING if engine P2P sync is enabled
```
```suggestion
		// Allow SYNCING and ACCEPTED if engine P2P sync is enabled
```
following the code order.

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -412,6 +431,26 @@ func (eq *EngineQueue) tryUpdateEngine(ctx context.Context) error {
+	return status == eth.ExecutionValid
+}
+
+// checkNewPayloadStatus checks returned status of engine_forkchoiceUpdatedV1 request for next unsafe payload.
```
```suggestion
// checkForkchoiceUpdatedStatus checks returned status of engine_forkchoiceUpdatedV1 request for next unsafe payload.
```

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -468,13 +506,17 @@ func (eq *EngineQueue) tryNextUnsafePayload(ctx context.Context) error {
 	}
 
-	eq.unsafeHead = ref
+	eq.engineSyncTarget = ref
```
```suggestion
        // engineSyncTarget should be updated only if the payload status is SYNCING or VALID
	eq.engineSyncTarget = ref
```

pcw109550 reviewed (changes requested):
We need to record when `engineSyncTarget` is modified, using `Metrics`. Refer `op-node/rollup/derive/pipeline.go:Metrics` interface.

ImTei committed:
Apply suggestions from code review

Rename engine p2p enabled flag
Return Step() if p2p sync is in progress
Fix some comments

ImTei committed:
Fix typo for test case

ImTei committed:
Record engineSyncTarget metric

ImTei committed:
Change isUnsafeSynced to isEngineSyncing

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -603,6 +655,9 @@ func (eq *EngineQueue) forceNextSafeAttributes(ctx context.Context) error {
 
 func (eq *EngineQueue) StartPayload(ctx context.Context, parent eth.L2BlockRef, attrs *eth.PayloadAttributes, updateSafe bool) (errType BlockInsertionErrType, err error) {
+	if eq.isEngineSyncing() {
+		return BlockInsertTemporaryErr, fmt.Errorf("engine is in progess of unsafe sync")
```
Engine will be always trying to run unsafe sync. In this case, we need to early return when unsafe block is advancing to enginesynctarget. So I think 
```suggestion
		return BlockInsertTemporaryErr, fmt.Errorf("unsafe block advancing to reach engineSyncTarget")
```
will be more descriptive.

ImTei committed:
Fix eq.Step() to process safe attributes

ImTei committed:
Return EngineSyncing error from eq.Step()

pcw109550 commented on op-node/rollup/derive/error.go:
```diff
@@ -96,3 +96,6 @@ var ErrCritical = NewCriticalError(nil)
 var NotEnoughData = errors.New("not enough data")
+
+// EngineSyncing implies that the execution engine is currently in progress of P2P sync.
+var EngineSyncing = errors.New("engine is syncing")
```
```suggestion
var EngineP2PSyncing = errors.New("engine is p2p syncing")
```
I suggest this change to distinguish between p2p sync and normal sync.

pcw109550 commented on op-node/rollup/driver/state.go:
```diff
@@ -290,7 +290,7 @@ func (s *Driver) eventLoop() {
 			err := s.derivation.Step(context.Background())
 			stepAttempts += 1 // count as attempt by default. We reset to 0 if we are making healthy progress.
-			if err == io.EOF {
+			if err == io.EOF || (err != nil && errors.Is(err, derive.EngineSyncing)) {
```
Is this equivalent to 

```suggestion
			if err == io.EOF || errors.Is(err, derive.EngineSyncing) {
```

??

ImTei committed:
Rename EngingSyncing err

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -725,6 +787,7 @@ func (eq *EngineQueue) Reset(ctx context.Context, _ eth.L1BlockRef, _ eth.System
 	}
 	eq.log.Debug("Reset engine queue", "safeHead", safe, "unsafe", unsafe, "safe_timestamp", safe.Time, "unsafe_timestamp", unsafe.Time, "l1Origin", l1Origin)
 	eq.unsafeHead = unsafe
+	eq.engineSyncTarget = unsafe
```
Is it okay to reset `engineSyncTarget` to `unsafe`? Any side effects?

pcw109550 commented:
At 
https://github.com/testinprod-io/optimism/blob/a6f4342434a7d2849f42be656ce204ccddfa68e3/op-node/rollup/derive/pipeline.go#L204-L212

now `Step` method may return `EngineP2PSyncing`. If this error is returned, is this considered a engine state failure? I think not.

```go
	// Now step the engine queue. It will pull earlier data as needed.
	if err := dp.eng.Step(ctx); err == io.EOF {
		// If every stage has returned io.EOF, try to advance the L1 Origin
		return dp.traversal.AdvanceL1Block(ctx)
	} else if err == EngineP2PSyncing {
		return err
	} else if err != nil {
		return fmt.Errorf("engine stage failed: %w", err)
	} else {
		return nil
	}
```

may more clear.

pcw109550 commented on op-node/rollup/driver/state.go:
```diff
@@ -290,7 +290,7 @@ func (s *Driver) eventLoop() {
 			stepAttempts += 1 // count as attempt by default. We reset to 0 if we are making healthy progress.
-			if err == io.EOF {
+			if err == io.EOF || (err != nil && errors.Is(err, derive.EngineP2PSyncing)) {
 				s.log.Debug("Derivation process went idle", "progress", s.derivation.Origin())
```
If `errors.Is(err, derive.EngineP2PSyncing)` is true, can we add more log messages to include error messages;
```suggestion
				s.log.Debug("Derivation process went idle", "progress", s.derivation.Origin(), "err", err)
```

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -221,10 +233,32 @@ func (eq *EngineQueue) SafeL2Head() eth.L2BlockRef {
+	if eq.unsafePayloads.Len() > 0 {
+		err := eq.tryNextUnsafePayload(ctx)
+		// EOF error means we can't process the next unsafe payload. Then we should process next safe attributes.
+		if err != io.EOF {
```
If `eq.syncCfg.EngineP2PEnabled` is true, `tryNextUnsafePayload` never returns `io.EOF` error, so control flow always falls in here. 

So why not 
```suggestion
		if eq.syncCfg.EngineP2PEnabled {
```

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -221,10 +233,32 @@ func (eq *EngineQueue) SafeL2Head() eth.L2BlockRef {
+	// Trying unsafe payload should be done before safe attributes
+	// It allows the unsafe head can move forward while the long-range consolidation is in progress.
+	if eq.unsafePayloads.Len() > 0 {
+		err := eq.tryNextUnsafePayload(ctx)
```
We can do it like
```suggestion
		if err := eq.tryNextUnsafePayload(ctx); err != io.EOF {
		     		     return err
		}
```

ImTei committed:
Explicit error handling

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -221,10 +233,31 @@ func (eq *EngineQueue) SafeL2Head() eth.L2BlockRef {
 		return eq.tryUpdateEngine(ctx)
 	}
+	// Trying unsafe payload should be done before safe attributes
+	// It allows the unsafe head can move forward while the long-range consolidation is in progress.
+	if eq.unsafePayloads.Len() > 0 {
+		if err := eq.tryNextUnsafePayload(ctx); err != io.EOF {
+			return err
+		}
+		// EOF error means we can't process the next unsafe payload. Then we should process next safe attributes.
+	}
```
```suggestion
        unsafePayloadEOF := false
	// It allows the unsafe head can move forward while the long-range consolidation is in progress.
	if eq.unsafePayloads.Len() > 0 {
		if err := eq.tryNextUnsafePayload(ctx); err != io.EOF {
			return err
		} else {
			unsafePayloadEOF = true
		}
		// EOF error means we can't process the next unsafe payload. Then we should process next safe attributes.
	}
```


If `err == io.EOF`, `tryNextUnsafePayload()` method is blocked for waiting new unsafe payloads. Suggestion to not ignore condition.

After safe block logic is done, we reach 
```go
	if outOfData || unsafePayloadEOF {
		return io.EOF
	} else {
		return 
	}
```

pcw109550 replied:
After offline discussion, we do not need this change. The code will eventually return `io.EOF` when reached upper code segment.

pcw109550 commented:
https://github.com/testinprod-io/optimism/blob/df0574ec294b83e8f9d4588a1bcc230196b5a72f/op-node/rollup/derive/engine_queue.go#L470-L481

We have changed to prioritize advancing unsafe blocks first instead of safe blocks(diff in  `Step()` method).

The meaning of `io.EOF` caughted at `stage.go` means engine queue had no information to derive, and even checked L1 ` dp.traversal.AdvanceL1Block(ctx)` to get more info.

 `tryNextUnsafePayload()` will never return `io.EOF` when p2p sync is enabled.

https://github.com/testinprod-io/optimism/blob/df0574ec294b83e8f9d4588a1bcc230196b5a72f/op-node/rollup/derive/engine_queue.go#L472-L482

Lets pick another `error` which indicates the message: we do not have enough `unsafePayload` to advance unsafe blocks. Now try to advance safe blocks.

pcw109550 commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -390,7 +420,7 @@ func (eq *EngineQueue) logSyncProgress(reason string) {
 func (eq *EngineQueue) tryUpdateEngine(ctx context.Context) error {
 	fc := eth.ForkchoiceState{
-		HeadBlockHash:      eq.unsafeHead.Hash,
+		HeadBlockHash:      eq.engineSyncTarget.Hash,
```
When this code is called, it is always `unsafeHead == engineSyncTarget`.
Lets check this invariant is satisfied. If not, lets log for this incorrect behavior

ImTei committed:
Add warning error
